[vm] Bug fix in dedup optimization when build deferred components.

lawangle opened the pull request:
`Dedup optimization` will remove duplicated Code object and keep one instance.
When building deferred components, the only Code object (e.x. as callee B) maybe
exist in different loading unit from the caller A. This will cause a rellocation
bug if A->B was a relative call.

Add a flag to indicate building with deferred components.
When working with dedup and deferred components, assignUnit first and
Code object from different unit is unequal.

Issues: https://github.com/dart-lang/sdk/issues/49372

Signed-off-by: fangzheyuan <fangzheyuan@bytedance.com>

lawangle commented:
This Pr fix one bug in building deferred component. I first found this error in https://github.com/dart-lang/sdk/issues/49372.
At first, i thought this bug only existed in older versions. But when i found out the real cause of the issue, i tnink this bug still exist in current master branch.

I will explain this bug.
`___TabsNonScrollableDemoState&State&SingleTickerProviderStateMixin&RestorationMixin._updateProperty) from Code([Optimized] __HomePageState&State&SingleTickerProviderStateMixin&RestorationMixin.registerForRestoration.listener)`
This error log shows than xxx.registerForRestoration.listener(caller) called xxx._updateProperty(callee) but callee could be
found in relocation phase.

__HomePageState is in unit4 and ___TabsNonScrollableDemoState is in another unit and this call entry's kind is relative call, that means caller and callee shouid in the same unit. But they are in the different units, so crash.

The cause is the dedup optimization. Actually 2.16.2's dedup. In 2.16.2, the dedup try to merge Code Object instead of instruciton. 
Before dedup, this call entry is from __HomePageState.xxx.registerForRestoration.listener to __HomePageState.xxx._updateProperty, relative call is ok. 
After dedup, __HomePageState.xxx._updateProperty is replaced with ___TabsNonScrollableDemoState.xxx and ___TabsNonScrollableDemoState is not exist in unit4.

mraleph commented:
Thanks for the contribution! Great catch. 

Initial code review comments are here: https://dart-review.googlesource.com/c/sdk/+/250491

lawangle committed:
[vm] Bug fix in dedup optimization when build deferred components.

Closes https://github.com/dart-lang/sdk/pull/49393

TEST=vm/dart{,_2}/regress_49372

Cq-Include-Trybots: luci.dart.try:vm-kernel-precomp-linux-debug-x64-try,vm-kernel-precomp-linux-product-x64-try
GitOrigin-RevId: c6306f96ae2bafa25c45e5c7a818c71eb72d8d12
Change-Id: Ie766b4c41fc61835767e62e3c463878e0493f812

Signed-off-by: 方哲源 <fangzheyuan@bytedance.com>

lawangle commented:
Change to  https://dart-review.googlesource.com/c/sdk/+/250720/
